main/lsp/LSPPreprocessor.cc

```diff
@@ -125,64 +89,14 @@ void LSPPreprocessor::mergeFileChanges(absl::Mutex &mtx, QueueState &state) {
-        earliestActiveEditVersion = ttgs.minVersion(committed, earliestActiveEditVersion);
-
-        // Avoid canceling if the currently-running slow path has already been canceled.
-        if (!epochManager.wasTypecheckingCanceled()) {
-            for (auto &msg : pendingRequests) {
-                if (msg->isNotification() && msg->method() == LSPMethod::SorbetWorkspaceEdit) {
-                    Timer timeit(logger, "tryCancelSlowPath");
-                    auto &params = get<unique_ptr<SorbetWorkspaceEditParams>>(msg->asNotification().params);
-                    auto combinedUpdates = ttgs.getCombinedUpdates(committed + 1, params->updates.versionEnd);
-                    // Cancel if combined updates end up taking the fast path, or if the new updates will just take the
-                    // slow path a second time when the current slow path finishes.
-                    if ((combinedUpdates.canTakeFastPath || !params->updates.canTakeFastPath) &&
-                        epochManager.tryCancelSlowPath(params->updates.versionEnd)) {
-                        if (combinedUpdates.canTakeFastPath) {
-                            logger->debug(
-                                "[Preprocessor] Canceling typechecking, as edits {} thru {} can take fast path.",
-                                combinedUpdates.versionStart, combinedUpdates.versionEnd);
-                        } else {
-                            logger->debug(
-                                "[Preprocessor] Canceling typechecking, as new edits {} thru {} will just take "
-                                "the slow path again.",
-                                params->updates.versionStart, params->updates.versionEnd);
-                            combinedUpdates.updatedGS = getTypecheckingGS();
-                        }
-                        combinedUpdates.cancellationExpected = params->updates.cancellationExpected;
-                        params->updates = move(combinedUpdates);
-                    }
-                    break;
-                } else if (!msg->isDelayable()) {
-                    // Message is not delayable, and is not an edit. Can't cancel the slow path.
-                    break;
-                }
-            }
-        }
-    }
-
-    // Prune history for all messages no longer in queue and no longer being typechecked.
-    ttgs.pruneBefore(earliestActiveEditVersion);
-}
-
-unique_ptr<core::GlobalState> LSPPreprocessor::getTypecheckingGS() const {
-    return ttgs.getGlobalState().deepCopy();
 }
 
 unique_ptr<LSPMessage> LSPPreprocessor::makeAndCommitWorkspaceEdit(unique_ptr<SorbetWorkspaceEditParams> params,
                                                                    unique_ptr<LSPMessage> oldMsg) {
-    ttgs.commitEdits(params->updates);
-    if (!params->updates.canTakeFastPath) {
-        params->updates.updatedGS = getTypecheckingGS();
+    for (auto &u : params->updates) {
+        fileContents[u->path()] = u;
```

Comment:
It does need to track the latest file contents so we can process diffs sent from clients. We're tracking `shared_ptr<core::File>`, which are immutable objects shared with `GlobalState`, so this is _not_ adding O(workspace size) memory overhead. It's adding O(size_of_pointer * # of files in workspace).

Reply:
Actually, I'm going to push an update; `fileContents` should only contain the contents of open files, not all files. My bad.